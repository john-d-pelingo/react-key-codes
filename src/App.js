/* eslint-disable react/prefer-stateless-function */

import React, { Component } from 'react';
import ReactGA from 'react-ga';

import logo from './logo.svg';
import './App.css';

class App extends Component {
    constructor(props) {
        super(props);

        ReactGA.initialize('UA-70753213-5');
        // This just needs to be called once since we have no routes in this case.
        ReactGA.pageview(window.location.pathname);
    }

    render() {
        return (
            <div className="App">
                <div className="App-header">
                    <img src={ logo } className="App-logo" alt="logo" />
                    <h2>Welcome to React</h2>
                </div>
                <p className="App-intro">
                    To get started, edit <code>src/App.js</code> and save to reload.
                </p>
            </div>
        );
    }
}

export default App;
