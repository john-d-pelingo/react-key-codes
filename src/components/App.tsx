import { css } from 'emotion'
import React from 'react'

import { keyCodes } from '../constants/keyCodes'
import { pageTitle } from '../constants/webApp'
import { KeyCode } from './KeyCode'

interface IAppState {
  blurred: boolean
  newKey: string
  newKeyCode: null | string
}

// TODO: use function component with hooks
export class App extends React.Component<{}, IAppState> {
  private appDomNode = React.createRef<HTMLDivElement>()

  constructor(props: {}) {
    super(props)

    /* istanbul ignore next */
    if (process.env.NODE_ENV !== 'test') {
      // eslint-disable-next-line global-require
      const ReactGA = require('react-ga')

      ReactGA.initialize('UA-70753213-5')
      // This just needs to be called once since we have no routes in this case.
      ReactGA.pageview(window.location.pathname)
    }
  }

  state = {
    blurred: false,
    newKey: '',
    newKeyCode: null,
  }

  componentDidMount() {
    // Focus the DOM element
    if (this.appDomNode.current) {
      this.appDomNode.current.focus()
    }
  }

  componentWillUpdate(_: Unused, nextState: IAppState) {
    if (nextState.newKeyCode && nextState.newKey) {
      document.title = `${nextState.newKeyCode} : ${nextState.newKey}`
    }
  }

  componentDidUpdate() {
    if (this.state.blurred) {
      this.setState({
        blurred: false,
      })
    }
  }

  handleBlur = () => {
    this.setState({
      blurred: true,
    })
  }

  handleClick = (event: React.MouseEvent<HTMLSpanElement>) => {
    event.preventDefault()

    this.setState(
      {
        newKeyCode: null,
      },
      () => {
        document.title = pageTitle
      },
    )
  }

  handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
    // Keep the original synthetic event around.
    // See https://fb.me/react-event-pooling for more information.
    event.persist()

    if (!event.metaKey) {
      event.preventDefault()
    }

    const newKeyCode =
      typeof event.which === 'number' ? event.which : event.keyCode

    this.setState({
      newKey: keyCodes[newKeyCode]
        ? keyCodes[newKeyCode]
        : event.key.toLowerCase(),
      newKeyCode: String(newKeyCode),
      // or event.which || event.keyCode || 0;
    })
  }

  render() {
    const { newKey, newKeyCode } = this.state

    return (
      // tabindex="0" allows elements besides links and form elements to receive keyboard focus
      // See http://webaim.org/techniques/keyboard/tabindex for more information
      <div
        className={appCss}
        tabIndex={0}
        // TODO: what does this do?
        onBlur={this.handleBlur}
        // TODO: maybe attach to window
        onKeyDown={this.handleKeyDown}
        ref={this.appDomNode}
      >
        {newKeyCode ? (
          <KeyCode
            keyCode={newKeyCode}
            keyText={newKey}
            handleClick={this.handleClick}
          />
        ) : (
          <span className="press-something">
            Press something in your keyboard
          </span>
        )}
      </div>
    )
  }
}

// -------
// STYLING
// -------

const appCss = css`
  display: table;
  height: 100%;
  margin: 0 auto;
  text-align: center;
  width: 100%;

  .press-something {
    display: table-cell;
    font-size: 50px;
    line-height: 55px;
    vertical-align: middle;
  }
`
