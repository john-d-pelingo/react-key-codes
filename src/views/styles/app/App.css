html, body, #root {
    height     : 100%;
    min-height : 400px;
}

.app {
    text-align : center;
    display    : table;
    height     : 100%;
    margin     : 0 auto;
}

.app .press-something {
    display        : table-cell;
    vertical-align : middle;
    font           : 50px/55px 'Raleway', 'Helvetica Neue', 'Helvetica', Arial, sans-serif;
}
