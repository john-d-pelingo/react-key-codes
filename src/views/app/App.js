/* eslint-disable jsx-a11y/no-noninteractive-tabindex, jsx-a11y/no-static-element-interactions */

import React from 'react';

import { KeyCode } from 'views/components';

import KEY_CODES from 'core/constants';

class App extends React.Component {
  constructor(props) {
    super(props);

    /* istanbul ignore next */
    if (process.env.NODE_ENV !== 'test') {
      // eslint-disable-next-line global-require
      const ReactGA = require('react-ga');

      ReactGA.initialize('UA-70753213-5');
      // This just needs to be called once since we have no routes in this case.
      ReactGA.pageview(window.location.pathname);
    }

    this.state = {
      blurred: false,
      defaultDocumentTitle: 'React Key Codes',
      id: 'app',
      newKey: '',
      newKeyCode: null,
      previousKeyCode: ''
    };

    // this._setStateBlurred = this._setStateBlurred.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
    this.handleClick = this.handleClick.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  componentDidMount() {
    this._setStateBlurred(false);
  }

  componentWillUpdate(nextProps, nextState) {
    if (nextState.newKeyCode && nextState.newKey) {
      document.title = `${ nextState.newKeyCode } : ${ nextState.newKey }`;
    } else {
      document.title = nextState.defaultDocumentTitle;
    }
  }

  componentDidUpdate() {
    if (this.state.blurred === true) {
      this._setStateBlurred(false);
    }
  }

  _setStateBlurred(blurred = false) {
    this.appDOMNode.focus();
    return this.setState(function () {
      return {
        blurred
      };
    });
  }

  handleBlur() {
    this.setState(function () {
      return {
        blurred: true
      };
    });
  }

  handleClick(event) {
    event.preventDefault();

    // document.title = this.state.defaultDocumentTitle;
    this.setState({
      newKeyCode: null,
      previousKeyCode: ''
    });
  }

  handleKeyDown(event) {
    // Keep the original synthetic event around.
    // See https://fb.me/react-event-pooling for more information.
    event.persist();

    if (!event.metaKey) {
      event.preventDefault();
    }

    const newKeyCode = (typeof event.which === 'number') ? event.which : event.keyCode;
    this.setState(function (prevState) {
      return {
        newKey: KEY_CODES[newKeyCode] ? KEY_CODES[newKeyCode] : event.key.toLowerCase(),
        newKeyCode,
        // or event.which || event.keyCode || 0;
        previousKeyCode: prevState.newKeyCode
      };
    });
  }

  render() {
    const { newKey, newKeyCode } = this.state;

    const conditionallyRenderKeyCode = () => {
      return newKeyCode ?
        (<KeyCode keyCode={ newKeyCode } keyText={ newKey } handleClick={ this.handleClick } />)
        : (<span className="press-something">Press something in your keyboard</span>);
    };

    return (
      // tabindex="0" allows elements besides links and form elements to receive keyboard focus.
      // See http://webaim.org/techniques/keyboard/tabindex for more information.
      <div className="app" id={ this.state.id } tabIndex="0" onBlur={ this.handleBlur } onKeyDown={ this.handleKeyDown } ref={ appDOMNode => { this.appDOMNode = appDOMNode; } }>
        { conditionallyRenderKeyCode() }
      </div>
    );
  }
}

export default App;
