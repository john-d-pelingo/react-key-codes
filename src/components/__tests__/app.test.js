import React from 'react';
import ReactDOM from 'react-dom';

import App from '../app';

describe('App Main Container', () => {
  let defaultState;

  beforeEach(() => {
    defaultState = {
      blurred: true,
      defaultDocumentTitle: 'React Key Codes',
      id: 'app',
      newKey: '',
      newKeyCode: null,
      previousKeyCode: ''
    };
  });

  describe.skip('Default suite', () => {
    it('should render without crashing', () => {
      const div = document.createElement('div');
      ReactDOM.render(<App />, div);
    });

    it('should render without throwing an error', function() {
      expect(
        shallow(<App />).contains(
          <span className="press-something">
            Press something in your keyboard
          </span>
        )
      ).toBe(true);
    });

    it('should be selectable by class ".app"', () => {
      expect(shallow(<App />).is('.app')).toBe(true);
    });

    it('should mount in a full DOM', () => {
      const wrapper = mount(<App />);

      expect(wrapper.find('.app').length).toBe(1);
      expect(wrapper.find('.key-code')).toHaveLength(0);
    });

    it('should render "Press something in your keyboard" with no keyCode provided', () => {
      expect(render(<App />).text()).toEqual(
        'Press something in your keyboard'
      );
    });
  });

  describe('Default state', () => {
    it('should have blurred set to false when mounted', () => {
      expect(mount(<App />).state()).toEqual({
        ...defaultState,
        blurred: false
      });
    });
  });

  describe('New state', () => {
    it('shouldn\'t render "Press something in your keyboard" with keyCode provided', () => {
      const newState = {
        ...defaultState,
        blurred: false,
        newKeyCode: 188
      };

      expect(
        shallow(<App />)
          .setState(newState)
          .text()
      ).toContain('KeyCode');
    });
  });

  describe('Simulation', () => {
    it('#blur: should have blurred set to false after a blur event occurs', () => {
      const wrapper = mount(<App />);
      const newState = { ...defaultState, blurred: false };

      expect(mount(<App />).state()).toEqual(newState);
      wrapper.find('.app').simulate('blur');
      expect(mount(<App />).state()).toEqual(newState);
    });

    it('#click: should show the default message when the key character is clicked', () => {
      const wrapper = mount(<App />);
      const newState = {
        ...defaultState,
        newKey: 'Tab',
        newKeyCode: 9,
        blurred: false
      };

      expect(wrapper.setState(newState).find('.press-something')).toHaveLength(
        0
      );
      wrapper.find('.app .key-code-text').simulate('click');
      expect(wrapper.find('.press-something')).toHaveLength(1);
    });

    it('#keyDown: should remove default message when a key is pressed', () => {
      describe('tab key', () => {
        const wrapper = mount(<App />);

        expect(wrapper.find('.press-something')).toHaveLength(1);
        wrapper.find('.app').simulate('keyDown', {
          key: 'Tab',
          keyCode: 9,
          metaKey: true,
          which: 9
        });
        expect(wrapper.find('.press-something')).toHaveLength(0);
      });

      describe('0 key', () => {
        describe('which', () => {
          const wrapper = mount(<App />);

          expect(wrapper.find('.press-something')).toHaveLength(1);
          wrapper
            .find('.app')
            .simulate('keyDown', { key: '0', metaKey: false, which: 48 });
          expect(wrapper.find('.press-something')).toHaveLength(0);
        });

        describe('keyCode', () => {
          const wrapper = mount(<App />);

          expect(wrapper.find('.press-something')).toHaveLength(1);
          wrapper
            .find('.app')
            .simulate('keyDown', { key: '0', keyCode: 48, metaKey: false });
          expect(wrapper.find('.press-something')).toHaveLength(0);
        });
      });

      describe('f19', () => {
        describe('which', () => {
          const wrapper = mount(<App />);

          expect(wrapper.find('.press-something')).toHaveLength(1);
          wrapper.find('.app').simulate('keyDown', {
            key: 'F13',
            keyCode: 130,
            metaKey: false,
            which: 130
          });
          expect(wrapper.find('.press-something')).toHaveLength(0);
        });
      });
    });
  });
});
