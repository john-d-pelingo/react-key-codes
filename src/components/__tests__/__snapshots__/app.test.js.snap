// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should with correct styles 1`] = `
.emotion-0 {
  display: table;
  height: 100%;
  margin: 0 auto;
  text-align: center;
  width: 100%;
}

.emotion-0 .press-something {
  display: table-cell;
  font-size: 50px;
  line-height: 55px;
  vertical-align: middle;
}

<App>
  <div
    className="emotion-0"
    onBlur={[Function]}
    onKeyDown={[Function]}
    tabIndex="0"
  >
    <span
      className="press-something"
    >
      Press something in your keyboard
    </span>
  </div>
</App>
`;
